torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -392,13 +392,15 @@ void Reducer::mark_variable_ready(VariableIndex index) {
   // Run finalizer function and kick off reduction for local_used_maps once the
   // final bucket was marked ready.
   if (next_bucket_ == buckets_.size()) {
-    // H2D from local_used_maps_ to local_used_maps_dev_
-    for (size_t i = 0; i < local_used_maps_.size(); i++) {
-      // We do async H2D to avoid the blocking overhead. The async copy and
-      // allreduce respect the current stream, so will be sequenced correctly.
-      local_used_maps_dev_[i].copy_(local_used_maps_[i], true);
+    if (!unused_parameters_.empty()) {
```

Comment:
This does not guarantee to be consistent across processes. Say all processes set `find_unused_parameters = true`, but some processes have unused parameters but other do not. In that case, some processes will never participating in this AllReduce, leading to either hang or wrong results. 

To know it deterministically, you might want to directly passing `find_unused_parameters` from Python to C++.

Reply:
also, does not need to allocate memory for local_used_maps_dev_ and local_used_maps if find_unused_parameters is false

Reply:
@zhaojuanmao Do you recommend to skip all use of `local_used_maps_dev_` and `local_used_maps` when `find_unused_param=False`? I think `local_used_maps_dev_` can be abandoned completely when `find_unused_param=False`, but `local_used_maps_` is still required in `Reducer::finalize_bucket_dense`.

Reply:
@sinannasir we can still keep the variables, but not allocate memory to them